### PR TITLE
Closes #6432: Bail out from beacon conditions

### LINKED ISSUE #6432: 3.16 - Bail out from beacon script if already processed or not acceptable screen size
## Context
As a user, I want WP Rocket to perform an accurate analysis of LCP/ATF images so that I get the best optimization possible.This issue has the goal to develop the orange part of this diagram:
⚠ The diagram mentions that, for not cached pages, the check about existing data should be bypassed. After discussion with Piotr, we don't really care so we can simplify the flow and check in all cases.
Untitled

**Must be done after #6431  so that it can be tested**

## Scope a solution 
- To implement this, we need to modify the beacon script from #6431 and add a check if we already have images related to the URL within the database. If yes or the status is failed, stop the script process.
- And also add a screen-size check within the script:
    - if one of the dimension is below the one of Pixel 4 (used in RUCSS as mobile: 393 x 830), bail out.
    - if one of the dimension is above the one of 1080p (1920x1080), bail out. 

For the first check, we could add an additional AJAX call that would trigger a method inside the AJAX controller (callback on the subscriber) that would check if there are any records for the current URL (using the query class).
- Add tests

## Acceptance Criteria
Bail-out if data is already available:
- Make sure there is a log in the beacon script in case it bails out becasue data is already available.
- Clear cache and ATF/LCP data. Make sure cache feature and LCP/ATF feature are enabled.
- Visit a page from desktop.
- Check in the DB that the LCP/ATF data is available for this page.
- Visit the page once again from desktop. Check in the dev tool console that the script about bailing out for data already there is printed out.
- Visit the page from mobile. Check that there is no log about bailing out for data laready there.
- Check in the DB that the mobile LCP/ATF data has been added.

Bail-out if not acceptable screen-size: (Use BrowserStack or Rocket-E2E)
- Make sure there is a log in the script if it bails out because of screen size.
- Clear cache and LCP/ATF data.
- Visit a page with a screen-size 392 x 1000. The script should bail out with the right log.
- Visit a page with a screen-size 1000 x 829. The script should bail out with the right log. 
- Visit a page with a screen-size 1000 x 1081. The script should bail out with the right log. 
- Visit a page with a screen-size 1921 x 1080. The script should bail out with the right log. 
- Visit a page with a screen-size 1920 x 1080. The script should complete and data must be available in the DB.
- Clear LCP/ATF data.
- Visit a page with a screen-size 393 x 830. The script should complete and data must be available in the DB.
- Clear LCP/ATF data.
- Visit a page with a screen-size 1000 x 1000. The script should complete and data must be available in the DB.

Check the screen-size filters:
- Repeat the above test about screen-size after applying different upper and lower limits with a filter:
    - Just set the upper limit: repeat the test (adapting the emulated screen-size).
    - Just set the lower limit: repeat the test (adapting the emulated screen-size).
- Validate that the filter is protected against unexpected inputs (null, empty, text, etc...)

##  Effort estimation:
Effort [S]

### PR DESCRIPTION
**Pull request by Miraeld**
# Description

Fixes #6432 

This PR modifies the beacon script to add a check for existing image data related to the URL in the database. If such data exists or the status is failed, the script process should be stopped. Additionally, a screen-size check should be added to the script to bail out under certain conditions.

### User documentation

This change does not directly impact users as it is related to the test setup and the beacon script.

### Technical documentation

The `add_lcp_data` and `check_lcp_data` methods in the `Controller.php` file are used to add and check LCP data respectively. The `main` function in the `lcp-beacon.js` file has been modified to make an AJAX call to check if there are any records for the current URL and to check the screen size. If the screen size is not within the acceptable range, the function bails out.

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies
N/A

## Risks
N/A

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Documentation

- [x] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [x] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [x] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitly.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] I listed the introduced external dependencies explicitly on the PR.
- [x] I validated the repo-specific guidelines from CONTRIBUTING.md.

----

## Acceptance Criteria Reminder

Bail-out if data is already available:
- [x] Make sure there is a log in the beacon script in case it bails out because data is already available.
- [x] Clear cache and ATF/LCP data. Make sure cache feature and LCP/ATF feature are enabled.
- [x] Visit a page from desktop.
- [x] Check in the DB that the LCP/ATF data is available for this page.
- [x] Visit the page once again from desktop. Check in the dev tool console that the script about bailing out for data already there is printed out.
- [x] Visit the page from mobile. Check that there is no log about bailing out for data already there.
- [x] Check in the DB that the mobile LCP/ATF data has been added.

Bail-out if not acceptable screen-size: (Use BrowserStack or Rocket-E2E)
- [x] Make sure there is a log in the script if it bails out because of screen size.
- [x] Clear cache and LCP/ATF data.
- [x] Visit a page with a screen-size 392 x 1000. The script should bail out with the right log.
- [x] Visit a page with a screen-size 1000 x 829. The script should bail out with the right log. 
- [x] Visit a page with a screen-size 1000 x 1081. The script should bail out with the right log. 
- [ ] Visit a page with a screen-size 1921 x 1080. The script should bail out with the right log. 
- [x] Visit a page with a screen-size 1920 x 1080. The script should complete and data must be available in the DB.
- [x] Clear LCP/ATF data.
- [ ] Visit a page with a screen-size 393 x 830. The script should complete and data must be available in the DB.
- [ ] Clear LCP/ATF data.
- [ ] Visit a page with a screen-size 1000 x 1000. The script should complete and data must be available in the DB.

Check the screen-size filters:
- [ ] Repeat the above test about screen-size after applying different upper and lower limits with a filter:
    - [ ] Just set the upper limit: repeat the test (adapting the emulated screen-size).
    - [ ] Just set the lower limit: repeat the test (adapting the emulated screen-size).
- [ ] Validate that the filter is protected against unexpected inputs (null, empty, text, etc...)

### CODE REVIEW

**Comment by remyperona** [comment]
Is this ready for review?

**Comment by Miraeld** [comment]
@Tabrisrp, yes it is ready. 
We keep it like this at the moment. The documentation related to these changes can be found [here](https://www.notion.so/wpmedia/Beacon-Auto-optimize-LCP-image-Auto-exclude-images-above-the-fold-from-LazyLoad-66f9411b46b9414082ac985b5cb83fb1?pvs=4#ec7f67b35eab4ee4956876042c855671).

File: assets/js/lcp-beacon.js [hunk 1]
```diff
@@ -68,6 +68,31 @@ function getImageUrlFromElement(element) {
...
 
 async function main() {
+	// AJAX call to check if there are any records for the current URL
+	const response = await fetch(rocket_lcp_data.ajax_url, {
+		method: "POST",
+		credentials: 'same-origin',
+		body: new URLSearchParams({
+			action: 'rocket_check_lcp',
+			url: rocket_lcp_data.url,
+			rocket_lcp_nonce: rocket_lcp_data.nonce
```
Comment (remyperona) [inline assets/js/lcp-beacon.js]: isn't it missing the is_mobile value?

File: assets/js/lcp-beacon.js [hunk 2]
```diff
@@ -68,6 +68,31 @@ function getImageUrlFromElement(element) {
 let performance_images = [];
 
 async function main() {
+	// AJAX call to check if there are any records for the current URL
+	const response = await fetch(rocket_lcp_data.ajax_url, {
+		method: "POST",
+		credentials: 'same-origin',
+		body: new URLSearchParams({
```
Comment (remyperona) [inline assets/js/lcp-beacon.js]: We use this here, and FormData on the other AJAX request, possibly we need to choose one and use the same for both?

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php [hunk 3]
```diff
@@ -91,4 +91,34 @@ public function add_lcp_data() {
...
+	public function check_lcp_data() {
+		check_ajax_referer( 'rocket_lcp', 'rocket_lcp_nonce' );
+
+		if ( ! $this->context->is_allowed() ) {
+			wp_send_json_error( false );
+			return;
+		}
+
+		$url       = isset( $_POST['url'] ) ? untrailingslashit( esc_url_raw( wp_unslash( $_POST['url'] ) ) ) : '';
+		$is_mobile = isset( $_POST['is_mobile'] ) ? filter_var( wp_unslash( $_POST['is_mobile'] ), FILTER_VALIDATE_BOOL ) : false;
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/AJAX/Controller.php]:
```suggestion
		$is_mobile = isset( $_POST['is_mobile'] ) ? filter_var( wp_unslash( $_POST['is_mobile'] ), FILTER_VALIDATE_BOOLEAN ) : false;
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by remyperona** [comment]
You have some conflicts to fix in the JS

**Review by Khadreal — APPROVED** [review]

**Review by MathieuLamiot — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
@Miraeld There are conflicts to be resolved. Then it should be merged ASAP to avoid further conflicts with upcoming changes on the beacon.

**Comment by Miraeld** [comment]
@Mai-Saad, there is no test plan available for this yet, 
However, I've tested this PR. 
The way I've tested => Open a page with different dimensions while checking the devconsole for results. 

Please find screenshots of the result as follow:
## Desktop
### Case 1
Opening a page in desktop mode while not meeting the size resolution requirements, we bail out. 
beacon-not-ok-desktop-size

### Case 2
Opening a page in desktop while meeting the size resolution, we process and get results.
beacon-ok-desktop-size

### Case 3
Opening a page in desktop while we already have saved data within the database.
becaon-data-already-exists

## Mobile
### Case 1
Opening a page in mobile mode while not meeting the size resolution requirements, we bail out. 
beacon-not-ok-mobile-size

### Case 2
Opening a page in mobile while meeting the size resolution, we process and get results.
beacon-ok-mobile-size

### Case 3
Opening a page in mobile while we already have saved data within the database.
beacon-data-already-exists-mobile


Please review this results for now and if everything's okay, we can move it to QA Done. 
Thank you

**Comment by MathieuLamiot** [comment]
@Miraeld Maybe we need to test the new filters and also that they are guarded against bad values? (null, [], string, etc.). If wrong value, then the default value should be applied.

**Comment by Miraeld** [comment]
True that, Here is the following tests: 

## Changing values of the filter 
### Settings new threshold: 
``` php
// Change the width threshold for the LCP beacon.
add_filter( 'rocket_lcp_width_threshold', function( $width_threshold, $is_mobile, $url ) {
	return $is_mobile ? 500 : 2500;
}, 10, 3 );

// Change the height threshold for the LCP beacon.
add_filter( 'rocket_lcp_height_threshold', function( $height_threshold, $is_mobile, $url ) {
	return $is_mobile ? 900 : 1200;
}, 10, 3 );
```
Screenshot 2024-04-25 at 15 52 08
Screenshot 2024-04-25 at 15 52 39

Thresholds are working perfectly. 

## Passing `null` & `[]` to filters
``` php
// Change the width threshold for the LCP beacon.
add_filter( 'rocket_lcp_width_threshold', function( $width_threshold, $is_mobile, $url ) {
	return [];
}, 10, 3 );

// Change the height threshold for the LCP beacon.
add_filter( 'rocket_lcp_height_threshold', function( $height_threshold, $is_mobile, $url ) {
	return null;
}, 10, 3 );
```
Result: 
Screenshot 2024-04-25 at 16 02 02
As values from the filter were not valid, we set back the default value.


Also please see https://github.com/wp-media/wp-rocket/pull/6593 for new threshold values modification

**Comment by MathieuLamiot** [comment]
Ok on my side 👍
